src/analyzer.ts

```diff
@@ -8,762 +8,153 @@
- */
-interface DocumentDescriptor {
-  /**
-   * The elements from the document.
-   */
-  elements: ElementDescriptor[];
-
-  /**
-   * The features from the document
-   */
-  features: FeatureDescriptor[];
-
-  /**
-   * The behaviors from the document
-   */
-  behaviors: BehaviorDescriptor[];
-
-  href?: string;
-
-  imports?: DocumentDescriptor[];
-
-  parsedScript?: estree.Program;
-
-  html?: ParsedImport;
-}
-
-/**
- * The metadata of an entire HTML document, in promises.
- */
-interface AnalyzedDocument {
-  /**
-   * The url of the document.
-   */
-  href: string;
-  /**
-   * The parsed representation of the doc. Use the `ast` property to get
-   * the full `parse5` ast.
-   */
-  htmlLoaded: Promise<ParsedImport>;
-
-  /**
-   * Resolves to the list of this Document's transitive import dependencies.
-   */
-  depsLoaded: Promise<string[]>;
-
-  /**
-   * The direct dependencies of the document.
-   */
-  depHrefs: string[];
-  /**
-   * Resolves to the list of this Document's import dependencies
-   */
-  metadataLoaded: Promise<DocumentDescriptor>;
-}
-
-/**
- * Options for `Analyzer.analzye`
- */
-interface LoadOptions {
-  /**
-   * Whether `annotate()` should be skipped.
-   */
-  noAnnotations?: boolean;
-  /**
-   * Content to resolve `href` to instead of loading from the file system.
-   */
-  content?: string;
-  /**
-   * Whether the generated descriptors should be cleaned of redundant data.
-   */
-  clean?: boolean;
-  /**
-   * `xhr` to use XMLHttpRequest.
-   * `fs` to use the local filesystem.
-   * `permissive` to use the local filesystem and return empty files when a
-   * path can't be found.
-   * Default is `fs` in node and `xhr` in the browser.
-   */
-  resolver?: string;
-  /**
-   * A predicate function that indicates which files should be ignored by
-   * the loader. By default all files not located under the dirname
-   * of `href` will be ignored.
-   */
-  filter?: (path:string)=> boolean;
-}
-
-/**
- * An Error extended with location metadata.
- */
-interface LocError extends Error{
-  location: {line: number; column: number};
-  ownerDocument: string;
+import * as path from 'path';
+import * as urlLib from 'url';
+
+import {CssParser} from './css/css-parser';
+import {EntityFinder} from './entity/entity-finder';
+import {HtmlDocument} from './html/html-document';
+import {HtmlImportFinder} from './html/html-import-finder';
+import {HtmlParser} from './html/html-parser';
+import {HtmlScriptFinder} from './html/html-script-finder';
+import {HtmlStyleFinder} from './html/html-style-finder';
+import {ElementFinder} from './javascript/javascript-element-finder';
+import {JavaScriptParser} from './javascript/javascript-parser';
+import {Document} from './parser/document';
+import {Parser} from './parser/parser';
+import {
+  Descriptor,
+  DocumentDescriptor,
+  ImportDescriptor,
+} from './ast/ast';
+import {UrlLoader} from './url-loader/url-loader';
+
+export interface AnalyzerInit {
```

Comment:
Style note: I find that, when there's only one such interface in a file, it's clearer to call it just `Options`. No need to stutter the name since this name is scoped to the current module, and `Options` is the more common name for this kind of object in the js world than `Init`.